overlord/hookstate/hookmgr.go

```diff
@@ -0,0 +1,84 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package hookstate implements the manager and state aspects responsible for
+// the running of hooks.
+package hookstate
+
+import (
+	"fmt"
+
+	"gopkg.in/tomb.v2"
+
+	"github.com/snapcore/snapd/hooks"
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+// HookManager is responsible for the maintenance of hooks in the system state.
+// It runs hooks when they're requested, assuming they're present in the given
+// snap. Otherwise they're skipped with no error.
+type HookManager struct {
+	state  *state.State
+	runner *state.TaskRunner
+}
+
+// Manager returns a new HookManager.
+func Manager(s *state.State) (*HookManager, error) {
+	runner := state.NewTaskRunner(s)
+	manager := &HookManager{
+		state:  s,
+		runner: runner,
+	}
+
+	runner.AddHandler("run-hook", manager.doRunHook, nil)
+
+	return manager, nil
+}
+
+// Ensure implements StateManager.Ensure.
+func (m *HookManager) Ensure() error {
+	m.runner.Ensure()
+	return nil
+}
+
+// Wait implements StateManager.Wait.
+func (m *HookManager) Wait() {
+	m.runner.Wait()
+}
+
+// Stop implements StateManager.Stop.
+func (m *HookManager) Stop() {
+	m.runner.Stop()
+}
+
+func (m *HookManager) doRunHook(task *state.Task, _ *tomb.Tomb) error {
+	task.State().Lock()
+	defer task.State().Unlock()
+
+	var hookRef hooks.HookRef
+	if err := task.Get("hook", &hookRef); err != nil {
+		return fmt.Errorf("failed to extract hook from task: %s", err)
+	}
+
+	if err := hooks.DispatchHook(hookRef); err != nil {
```

Comment:
will DispatchHook be synchronous or asynchronous? (we are already in a go routine here because of TaskRunner so the former would be ok) either way it may need to take the Tomb to be interruptable, and if it's synchronous we should release the state lock during/before it